File: Problem_Solving/leetcode/Sequential(Non-Linear)_data_structure/Graph/BFS/BFS.py
# 큐를 이용한 반복 구조

# 모든 인접 Edge를 추출하고 도착점인 Vertex를 큐에 삽입
def iterative_bfs(start_v):
    discovered = [start_v]
    queue = [start_v]   # 최적화는 위해 dequeue도괜찮다.
    while queue:
        v = queue.pop(0)
        # 인접 Edge를 추출해 다음 도착 Vertex 삽입
        for w in graph[v]:
            # 가본적이 없는 vertex라면
            if w not in discovered:
                discovered.append(w)
                queue.append(w)
    return discovered

# 재귀함수로는 BFS를 구현할 수 없다.
    
    
if __name__ == "__main__":
    graph = {
        1: [2, 3, 4],
        2: [5],
        3: [5],
        4: [],
        5: [6, 7],
        6: [],
        7: [3]
    }

File: Problem_Solving/leetcode/Sequential(Non-Linear)_data_structure/Graph/BFS/BFS.cpp
#include <vector>
#include <queue>

// 큐를 이용한 반복구조
std::vector<int> iterative_bfs(int v, std::vector<std::vector<int>>& graph)
{
    std::vector<int> visited = {v};
    std::queue<int> queue;
    queue.push(v);

    while (!queue.empty())
    {
        v = queue.front();
        queue.pop();
        // 인접 Edge를 추출해 다음 vertex 삽입
        for (auto w : graph[v])
        {
            // 가본적 없는 vertex라면
            if (std::find(visited.begin(), visited.end(), w) == visited.end())
            {
                visited.push_back(v);
                queue.push(v);
            }
        }
    }

    return visited;
}

// 재귀함수로는 BFS를 구현할 수 없다.

int main()
{
    std::vector<std::vector<int>> graph =
    {
        {2, 3, 4},
        {5},
        {},
        {6,7},
        {},
        {3}
    };

    return 0;
}